[![DOI](https://zenodo.org/badge/688164096.svg)](https://zenodo.org/doi/10.5281/zenodo.10640541)


# ğŸ§¬ Project Overview

Welcome to the code repository for our paper:

ğŸ“„ Alternative Splicing Decouples Local from Global PRC2 Activity

This repository houses the code used to generate the figures presented in the paper. The code is organized into `R` quarto notebooks, with each notebook dedicated to one of the seven figures in the paper.

For easy access to each analysis report, visit our main homepage with direct links to the analysis reports:

ğŸ”— [Main Homepage](https://htmlpreview.github.io/?https://github.com/Ni-Ar/SUZ12_AS/blob/main/_Code/index.html)

Feel free to explore the code and reproduce our findings! ğŸŒŸ

If you use our work, please cite us appropriately 

# ğŸ“– Citation

[BioRÏ‡iv](https://www.biorxiv.org/content/10.1101/2023.04.30.538846v1)

> Arecco, N.â€ , Mocavini, I.â€ , Blanco, E., BallarÃ©, C., Bonnal, S., Irimia, M., & Di Croce, L. (2023). 
> *Alternative Splicing Decouples Local from Global PRC2 Activity* **BioRÏ‡iv**.
> DOI: 10.1101/2023.04.30.538846

# Related links

- ğŸ¦ [BioRÏ‡iv twitter thread](https://twitter.com/NccAcc/status/1653308468609073153)
- ğŸ§  [Alternative splicing database](https://vastdb.crg.eu/wiki/Main_Page) 
- ğŸ’¡ ChIP-seq and RNA-seq fastq files ([NCBI GEO](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE223666))
- ğŸ’ª Mass spectrometry raw files (PRIDE: [PXD046640](https://www.ebi.ac.uk/pride/archive/projects/PXD046640/), [PXD046660](https://www.ebi.ac.uk/pride/archive/projects/PXD046660/), [PXD046742]( https://www.ebi.ac.uk/pride/archive/projects/PXD046742/))
- ğŸ— Colabfold PRC2 structural predictions ([GitHub repo](https://github.com/Ni-Ar/colabfold_crg))
- ğŸ”¬Immunofluorescence microscopy czi files ([Mendeley data](https://data.mendeley.com/datasets/sb72467mrf/1))

